python/app/input_widget.py

```diff
@@ -227,20 +228,364 @@ def keyPressEvent(self, event):
+            else:
+                super(PythonInputWidget, self).keyPressEvent(event)
+        elif event.key() == QtCore.Qt.Key_Slash:
+            self.block_comment_selection()
+            event.accept()
+        elif event.key() == QtCore.Qt.Key_Backtab:
+            self.unindent()
             event.accept()
         elif event.key() == QtCore.Qt.Key_Tab:
             # intercept the tab key and insert 4 spaces
-            self.insertPlainText("    ")
+            self.indent()
             event.accept()
         else:
             super(PythonInputWidget, self).keyPressEvent(event)
 
+    def remove_character_indentation(self):
+        """
+        Attempts to remove a single indentation block if there is no selection and the cursor is in the indentation
+        part of the line. It returns True if we unindented and False if we didn't.
+        :return: bool
+        """
+        cur = self.textCursor()
+        if not cur.hasSelection():
+            user_cur_pos = cur.positionInBlock()
+            line = cur.block().text()
+
+            # Now separate out the indentation from the rest of the line and figure out how many spaces it is.
+            indentation, rest_of_line = self._split_indentation(line)
+
+            # Te current user cursor position is within the indentation.
+            if user_cur_pos <= len(indentation) and len(indentation) != 0:
+                self.unindent()
+                return True
+        return False
+
+    def add_new_line(self):
+        """
+        Adds a new line from the cursor position.
+        The new line will be indented at the same level as the previous line.
+        :return: None
+        """
+        cur = self.textCursor()
+
+        # Start the undo block
+        cur.beginEditBlock()
+
+        cur_pos, anchor, start, end = self._get_cursor_positions(cur)
+
+        # We only want to match the indentation of the up most selected line.
+        cur.setPosition(start)
+
+        # Get the text for the current line, so that we can figure out the current line's indentation.
+        line = cur.block().text()
+
+        # Now separate out the indentation from the rest of the line and figure out how many spaces it is.
+        indentation, rest_of_line = self._split_indentation(line)
+        n_spaces = self._get_indentation_length(indentation)
+
+        # We should check the current position in the line, if it sits somewhere in the indentation then match that.
+        cur_pos_in_block = cur.positionInBlock()
+        # Note don't use n_spaces in this calculation as tabs might have been removed.
+        if not self.textCursor().hasSelection() and cur_pos_in_block < len(indentation):
+            n_spaces = cur_pos_in_block
+
+        # Check if the current line has a `:` at the end. Also account for any white spaces after it.
+        # If we find one, then the next line should auto indent four spaces.
+        check_for_colon = re.compile(r":[ \t]*$")
```

Comment:
Have you considered a line ending with a comment?

For example, this would not work:
`def allo(): #allo`

Reply:
Maybe you should check into using the AST module for that one.

Reply:
I didn't know about the ast module, however, I've updated the regex to handle comments. Good catch I'd forgotten about that.
If you still think it's worth investigating the ast module I can do.

Reply:
+ I've added a test to include comments at the end of the line with new line and `:`